File: project/src/main/java/com/expedia/ui/javademo/ItemVH.java
package com.expedia.ui.javademo;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.expedia.ui.javademo.interfaces.HolderAdapterBridge;
import com.expedia.ui.javademo.interfaces.VHClickable;

/**
 * Created by nbirla on 15/11/17.
 */

public abstract class ItemVH<V> extends RecyclerView.ViewHolder implements View.OnClickListener {

    private View parent;
    protected HolderAdapterBridge adapterBridge;
    private FeedItem feedItem;
    private VHClickable clickCallback;

    public ItemVH(View itemView) {
        super(itemView);
        parent = itemView;
    }

    public View getParent() {
        return parent;
    }

    public void bindFeedItem(FeedItem <V> feedItem) {
        this.feedItem = feedItem;
        bindData(feedItem.getBindingData());
    }

    protected abstract void bindData(V v);

    protected FeedItem<V> getFeedItem(){
        return feedItem;
    }

    @Override
    public void onClick(View v) {
        if(clickCallback != null){
            clickCallback.onViewHolderClicked(this, v);
        }
    }

    public void setVHClickCallback(VHClickable callback) {
        this.clickCallback = callback;
    }

    public VHClickable getVHClickable() {
        return clickCallback;
    }

    public void setGLAdapterBridge(HolderAdapterBridge adapterBridge) {
        this.adapterBridge = adapterBridge;
    }

    protected void notifyVHChanged() {
        if(adapterBridge != null) {
            adapterBridge.notifyVHChanged(getAdapterPosition());
        }
    }
}

File: project/src/main/java/com/expedia/ui/recyclerview/ItemVH.kt
package com.expedia.ui.recyclerview

import android.support.v7.widget.RecyclerView
import android.view.View
import butterknife.ButterKnife
import com.expedia.ui.recyclerview.interfaces.VHClickable
import com.expedia.ui.recyclerview.interfaces.HolderAdapterBridge

/**
 * Created by nbirla on 15/11/17.
 */
abstract class ItemVH<V>(val root: View) : RecyclerView.ViewHolder(root), View.OnClickListener {

    protected var adapterBridge: HolderAdapterBridge? = null
    private var feedItem: FeedItem<V>? = null
    private var clickCallback: VHClickable? = null

    init {
        ButterKnife.inject(this, root);
    }

    fun getParent(): View {
        return root
    }

    fun bindFeedItem(feedItem: FeedItem<*>) {
        val feedItem = FeedItem<V>(feedItem as FeedItem<V>)
        this.feedItem = feedItem
        bindData(feedItem.getBindingData())
    }

    protected abstract fun bindData(v: V)

    protected fun getFeedItem(): FeedItem<V>? {
        return feedItem
    }

    override fun onClick(v: View) {
        clickCallback!!.onViewHolderClicked(this, v)
    }

    fun setVHClickCallback(callback: VHClickable) {
        this.clickCallback = callback
    }

    fun getVHClickable(): VHClickable? {
        return clickCallback
    }

    fun setGLAdapterBridge(adapterBridge: HolderAdapterBridge) {
        this.adapterBridge = adapterBridge
    }

    protected fun notifyVHChanged() {
        adapterBridge!!.notifyVHChanged(adapterPosition)
    }
}